File: core/src/main/java/com/graphhopper/util/BitUtil.java
/*
 *  Licensed to GraphHopper GmbH under one or more contributor
 *  license agreements. See the NOTICE file distributed with this work for
 *  additional information regarding copyright ownership.
 *
 *  GraphHopper GmbH licenses this file to you under the Apache License,
 *  Version 2.0 (the "License"); you may not use this file except in
 *  compliance with the License. You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package com.graphhopper.util;

import com.graphhopper.storage.IntsRef;

/**
 * LITTLE endianness is default for GraphHopper and most microprocessors.
 *
 * @author Peter Karich
 */
public class BitUtil {
    public static final BitUtil LITTLE = new BitUtil();

    public final double toDouble(byte[] bytes) {
        return toDouble(bytes, 0);
    }

    public final double toDouble(byte[] bytes, int offset) {
        return Double.longBitsToDouble(toLong(bytes, offset));
    }

    public final byte[] fromDouble(double value) {
        byte[] bytes = new byte[8];
        fromDouble(bytes, value, 0);
        return bytes;
    }

    public final void fromDouble(byte[] bytes, double value) {
        fromDouble(bytes, value, 0);
    }

    public final void fromDouble(byte[] bytes, double value, int offset) {
        fromLong(bytes, Double.doubleToRawLongBits(value), offset);
    }

    public final float toFloat(byte[] bytes) {
        return toFloat(bytes, 0);
    }

    public final float toFloat(byte[] bytes, int offset) {
        return Float.intBitsToFloat(toInt(bytes, offset));
    }

    public final byte[] fromFloat(float value) {
        byte[] bytes = new byte[4];
        fromFloat(bytes, value, 0);
        return bytes;
    }

    public final void fromFloat(byte[] bytes, float value) {
        fromFloat(bytes, value, 0);
    }

    public final void fromFloat(byte[] bytes, float value, int offset) {
        fromInt(bytes, Float.floatToRawIntBits(value), offset);
    }

    public final short toShort(byte[] b) {
        return toShort(b, 0);
    }

    public final short toShort(byte[] b, int offset) {
        return (short) ((b[offset + 1] & 0xFF) << 8 | (b[offset] & 0xFF));
    }

    public final int toInt(byte[] b) {
        return toInt(b, 0);
    }

    public final int toInt(byte[] b, int offset) {
        return (b[offset + 3] & 0xFF) << 24 | (b[offset + 2] & 0xFF) << 16
                | (b[offset + 1] & 0xFF) << 8 | (b[offset] & 0xFF);
    }

    public final byte[] fromInt(int value) {
        byte[] bytes = new byte[4];
        fromInt(bytes, value, 0);
        return bytes;
    }

    public final void fromInt(byte[] bytes, int value) {
        fromInt(bytes, value, 0);
    }

    public final byte[] fromShort(short value) {
        byte[] bytes = new byte[4];
        fromShort(bytes, value, 0);
        return bytes;
    }

    public final void fromShort(byte[] bytes, short value) {
        fromShort(bytes, value, 0);
    }

    public void fromShort(byte[] bytes, short value, int offset) {
        bytes[offset + 1] = (byte) (value >>> 8);
        bytes[offset] = (byte) (value);
    }

    public final void fromInt(byte[] bytes, int value, int offset) {
        bytes[offset + 3] = (byte) (value >>> 24);
        bytes[offset + 2] = (byte) (value >>> 16);
        bytes[offset + 1] = (byte) (value >>> 8);
        bytes[offset] = (byte) (value);
    }

    public final long toLong(byte[] b) {
        return toLong(b, 0);
    }

    public final long toLong(int int0, int int1) {
        return ((long) int1 << 32) | (int0 & 0xFFFFFFFFL);
    }

    public final long toLong(byte[] b, int offset) {
        return ((long) toInt(b, offset + 4) << 32) | (toInt(b, offset) & 0xFFFFFFFFL);
    }

    public final byte[] fromLong(long value) {
        byte[] bytes = new byte[8];
        fromLong(bytes, value, 0);
        return bytes;
    }

    public final void fromLong(byte[] bytes, long value) {
        fromLong(bytes, value, 0);
    }

    public final void fromLong(byte[] bytes, long value, int offset) {
        bytes[offset + 7] = (byte) (value >> 56);
        bytes[offset + 6] = (byte) (value >> 48);
        bytes[offset + 5] = (byte) (value >> 40);
        bytes[offset + 4] = (byte) (value >> 32);
        bytes[offset + 3] = (byte) (value >> 24);
        bytes[offset + 2] = (byte) (value >> 16);
        bytes[offset + 1] = (byte) (value >> 8);
        bytes[offset] = (byte) (value);
    }

    public byte[] fromBitString(String str) {
        // no need for performance or memory tuning ...
        int strLen = str.length();
        int bLen = str.length() / 8;
        if (strLen % 8 != 0)
            bLen++;

        byte[] bytes = new byte[bLen];
        int charI = 0;
        for (int b = bLen - 1; b >= 0; b--) {
            byte res = 0;
            for (int i = 0; i < 8; i++) {
                res <<= 1;
                if (charI < strLen && str.charAt(charI) != '0')
                    res |= 1;

                charI++;
            }
            bytes[b] = res;
        }
        return bytes;
    }

    public final String toBitString(IntsRef intsRef) {
        StringBuilder str = new StringBuilder();
        for (int ints : intsRef.ints) {
            str.append(toBitString(ints, 32));
        }
        return str.toString();
    }

    /**
     * Similar to Long.toBinaryString
     */
    public final String toBitString(long value) {
        return toBitString(value, 64);
    }

    public String toLastBitString(long value, int bits) {
        StringBuilder sb = new StringBuilder(bits);
        long lastBit = 1L << bits - 1;
        for (int i = 0; i < bits; i++) {
            if ((value & lastBit) == 0)
                sb.append('0');
            else
                sb.append('1');

            value <<= 1;
        }
        return sb.toString();
    }

    /**
     * Higher order bits comes first in the returned string.
     * <p>
     *
     * @param bits how many bits should be returned.
     */
    public String toBitString(long value, int bits) {
        StringBuilder sb = new StringBuilder(bits);
        long lastBit = 1L << 63;
        for (int i = 0; i < bits; i++) {
            if ((value & lastBit) == 0)
                sb.append('0');
            else
                sb.append('1');

            value <<= 1;
        }
        return sb.toString();
    }

    /**
     * Higher order bits comes first in the returned string.
     */
    public String toBitString(byte[] bytes) {
        StringBuilder sb = new StringBuilder(bytes.length * 8);
        byte lastBit = (byte) (1 << 7);
        for (int bIndex = bytes.length - 1; bIndex >= 0; bIndex--) {
            byte b = bytes[bIndex];
            for (int i = 0; i < 8; i++) {
                if ((b & lastBit) == 0)
                    sb.append('0');
                else
                    sb.append('1');

                b <<= 1;
            }
        }
        return sb.toString();
    }

    public final int getIntLow(long longValue) {
        return (int) (longValue & 0xFFFFFFFFL);
    }

    public final int getIntHigh(long longValue) {
        return (int) (longValue >> 32);
    }

    public final long combineIntsToLong(int intLow, int intHigh) {
        return ((long) intHigh << 32) | (intLow & 0xFFFFFFFFL);
    }
}

File: core/src/main/kotlin/com/graphhopper/util/BitUtil.kt
/*
 *  Licensed to GraphHopper GmbH under one or more contributor
 *  license agreements. See the NOTICE file distributed with this work for
 *  additional information regarding copyright ownership.
 *
 *  GraphHopper GmbH licenses this file to you under the Apache License,
 *  Version 2.0 (the "License"); you may not use this file except in
 *  compliance with the License. You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package com.graphhopper.util

/**
 * LITTLE endianness is default for GraphHopper and most microprocessors.
 *
 * @author Peter Karich
 */
class BitUtil {
    fun toDouble(bytes: ByteArray): Double = toDouble(bytes, 0)

    fun toDouble(bytes: ByteArray, offset: Int): Double = Double.fromBits(toLong(bytes, offset))

    fun fromDouble(value: Double): ByteArray {
        val bytes = ByteArray(8)
        fromDouble(bytes, value, 0)
        return bytes
    }

    fun fromDouble(bytes: ByteArray, value: Double) {
        fromDouble(bytes, value, 0)
    }

    fun fromDouble(bytes: ByteArray, value: Double, offset: Int) {
        fromLong(bytes, value.toRawBits(), offset)
    }

    fun toFloat(bytes: ByteArray): Float = toFloat(bytes, 0)

    fun toFloat(bytes: ByteArray, offset: Int): Float = Float.fromBits(toInt(bytes, offset))

    fun fromFloat(value: Float): ByteArray {
        val bytes = ByteArray(4)
        fromFloat(bytes, value, 0)
        return bytes
    }

    fun fromFloat(bytes: ByteArray, value: Float) {
        fromFloat(bytes, value, 0)
    }

    fun fromFloat(bytes: ByteArray, value: Float, offset: Int) {
        fromInt(bytes, value.toRawBits(), offset)
    }

    fun toShort(b: ByteArray): Short = toShort(b, 0)

    fun toShort(b: ByteArray, offset: Int): Short =
        ((b[offset + 1].toInt() and 0xFF shl 8) or (b[offset].toInt() and 0xFF)).toShort()

    fun toInt(b: ByteArray): Int = toInt(b, 0)

    fun toInt(b: ByteArray, offset: Int): Int =
        (b[offset + 3].toInt() and 0xFF shl 24) or (b[offset + 2].toInt() and 0xFF shl 16) or
                (b[offset + 1].toInt() and 0xFF shl 8) or (b[offset].toInt() and 0xFF)

    fun toUInt3(b: ByteArray, offset: Int): Int =
        (b[offset + 2].toInt() and 0xFF shl 16) or (b[offset + 1].toInt() and 0xFF shl 8) or (b[offset].toInt() and 0xFF)

    fun fromInt(value: Int): ByteArray {
        val bytes = ByteArray(4)
        fromInt(bytes, value, 0)
        return bytes
    }

    fun fromInt(bytes: ByteArray, value: Int) {
        fromInt(bytes, value, 0)
    }

    fun fromShort(value: Short): ByteArray {
        val bytes = ByteArray(4)
        fromShort(bytes, value, 0)
        return bytes
    }

    fun fromShort(bytes: ByteArray, value: Short) {
        fromShort(bytes, value, 0)
    }

    fun fromShort(bytes: ByteArray, value: Short, offset: Int) {
        bytes[offset + 1] = (value.toInt() ushr 8).toByte()
        bytes[offset] = value.toByte()
    }

    fun fromInt(bytes: ByteArray, value: Int, offset: Int) {
        bytes[offset + 3] = (value ushr 24).toByte()
        bytes[offset + 2] = (value ushr 16).toByte()
        bytes[offset + 1] = (value ushr 8).toByte()
        bytes[offset] = value.toByte()
    }

    /**
     * Note, currently value with higher bits set (like for a negative value) won't throw an exception at this level.
     */
    fun fromUInt3(bytes: ByteArray, value: Int, offset: Int) {
        bytes[offset + 2] = (value ushr 16).toByte()
        bytes[offset + 1] = (value ushr 8).toByte()
        bytes[offset] = value.toByte()
    }

    /**
     * See the counterpart [fromLong]
     */
    fun toLong(b: ByteArray): Long = toLong(b, 0)

    fun toLong(intLow: Int, intHigh: Int): Long =
        (intHigh.toLong() shl 32) or (intLow.toLong() and 0xFFFF_FFFFL)

    fun toLong(b: ByteArray, offset: Int): Long =
        (toInt(b, offset + 4).toLong() shl 32) or (toInt(b, offset).toLong() and 0xFFFF_FFFFL)

    fun fromLong(value: Long): ByteArray {
        val bytes = ByteArray(8)
        fromLong(bytes, value, 0)
        return bytes
    }

    fun fromLong(bytes: ByteArray, value: Long) {
        fromLong(bytes, value, 0)
    }

    fun fromLong(bytes: ByteArray, value: Long, offset: Int) {
        bytes[offset + 7] = (value shr 56).toByte()
        bytes[offset + 6] = (value shr 48).toByte()
        bytes[offset + 5] = (value shr 40).toByte()
        bytes[offset + 4] = (value shr 32).toByte()
        bytes[offset + 3] = (value shr 24).toByte()
        bytes[offset + 2] = (value shr 16).toByte()
        bytes[offset + 1] = (value shr 8).toByte()
        bytes[offset] = value.toByte()
    }

    fun fromBitString(str: String): ByteArray {
        // no need for performance or memory tuning ...
        val strLen = str.length
        var bLen = str.length / 8
        if (strLen % 8 != 0)
            bLen++

        val bytes = ByteArray(bLen)
        var charI = 0
        for (b in bLen - 1 downTo 0) {
            var res = 0
            for (i in 0 until 8) {
                res = res shl 1
                if (charI < strLen && str[charI] != '0')
                    res = res or 1

                charI++
            }
            bytes[b] = res.toByte()
        }
        return bytes
    }

    /**
     * Similar to Long.toBinaryString
     */
    fun toBitString(value: Long): String = toBitString(value, 64)

    fun toLastBitString(value: Long, bits: Int): String {
        val sb = StringBuilder(bits)
        val lastBit = 1L shl (bits - 1)
        var v = value
        for (i in 0 until bits) {
            if (v and lastBit == 0L)
                sb.append('0')
            else
                sb.append('1')

            v = v shl 1
        }
        return sb.toString()
    }

    /**
     * Higher order bits comes first in the returned string.
     *
     * @param bits how many bits should be returned.
     */
    fun toBitString(value: Long, bits: Int): String {
        val sb = StringBuilder(bits)
        val lastBit = 1L shl 63
        var v = value
        for (i in 0 until bits) {
            if (v and lastBit == 0L)
                sb.append('0')
            else
                sb.append('1')

            v = v shl 1
        }
        return sb.toString()
    }

    /**
     * Higher order bits comes first in the returned string.
     */
    fun toBitString(bytes: ByteArray): String {
        val sb = StringBuilder(bytes.size * 8)
        val lastBit = (1 shl 7).toByte()
        for (bIndex in bytes.size - 1 downTo 0) {
            var b = bytes[bIndex]
            for (i in 0 until 8) {
                if (b.toInt() and lastBit.toInt() == 0)
                    sb.append('0')
                else
                    sb.append('1')

                b = (b.toInt() shl 1).toByte()
            }
        }
        return sb.toString()
    }

    fun getIntLow(longValue: Long): Int = (longValue and 0xFFFF_FFFFL).toInt()

    fun getIntHigh(longValue: Long): Int = (longValue shr 32).toInt()

    companion object {
        @JvmField
        val LITTLE = BitUtil()

        @JvmStatic
        fun countBitValue(maxTurnCosts: Int): Int {
            if (maxTurnCosts < 0)
                throw IllegalArgumentException("maxTurnCosts cannot be negative $maxTurnCosts")

            var v = maxTurnCosts
            var counter = 0
            while (v > 0) {
                v = v shr 1
                counter++
            }
            return counter
        }

        /**
         * Converts the specified long into a signed int ('reverse' method for Integer.toUnsignedLong).
         */
        @JvmStatic
        fun toSignedInt(x: Long): Int = x.toInt()
    }
}
